### PR TITLE
Wave quality improvement

### PR DESCRIPTION
**Pull request by huwb**
Expose option to double sample rate when generating final waves. This unfortunately requires higher res LOD data (1024 works well) because this effectively halves the range of detail and now at res=384 there is very little resolution left in the waves.

As previously, I'm finding higher LOD data resolution is required to get high quality results. To increase the default resolution, we'd likely need to break out LOD resolutions on a per-LOD basis as 1024x1024x8xTexelBytes x LodDataCount is a lot of video memory. For now i've added a tooltip that warns about this memory usage and labels the setting as preview.

Before [video](https://fr2ygz2xuzx8urqegqxn7uysnqr-my.sharepoint.com/:v:/g/personal/huw_bowles_waveharmonic_com/EScsHKPw375Aj1cClLx7ZJQB4vvtR-HGStAyczBSGThzNQ?e=MjWYkm)
After [video](https://fr2ygz2xuzx8urqegqxn7uysnqr-my.sharepoint.com/:v:/g/personal/huw_bowles_waveharmonic_com/Ebdyzv4msD5PmvdlERb5V7wBFB4ZocE6qz7GNMtQMotLrA?e=vaisyw)

The before/after shows some change in apparent wave amplitude, I believe this is a bad result of undersampling the waves rather than something being scaled wrong. I'll think about if this can be proven, but in any case I believe this can be merged.

### CODE REVIEW

File: crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs [hunk 1]
```diff
@@ -23,6 +23,12 @@ public partial class SimSettingsAnimatedWaves : SimSettingsBase
 
         public const string k_HelpURL = Internal.Constants.HELP_URL_BASE_USER + "ocean-simulation.html" + Internal.Constants.HELP_URL_RP + "#animated-waves-settings";
 
+        [Tooltip("PREVIEW: Set this to 2 to improve wave quality. In some cases like flowing rivers this can make a substantial difference to visual stability."
+            + "\n\nWe recommend doubling the Lod Data Resolution on the OceanRenderer component to preserve detail after making this change, but note that this will "
+            + "consume 4x more video memory until we are able to optimise data usage further, so apply this change with caution."), SerializeField]
+        float _waveResolutionMultiplier = 1f;
```
Comment (daleeidd) [inline crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs:29]: Should this be a range and have protection against divide-by-zero?

**Review by daleeidd — CHANGES_REQUESTED** [review]
Looks good. Just a small change needed.

File: crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs (hunk 1 above)
Comment (huwb) [reply crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs:30]: Yeah good shout done now

File: crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs [hunk 2]
```diff
@@ -23,6 +23,13 @@ public partial class SimSettingsAnimatedWaves : SimSettingsBase
 
         public const string k_HelpURL = Internal.Constants.HELP_URL_BASE_USER + "ocean-simulation.html" + Internal.Constants.HELP_URL_RP + "#animated-waves-settings";
 
+        [Tooltip("PREVIEW: Set this to 2 to improve wave quality. In some cases like flowing rivers this can make a substantial difference to visual stability."
+            + "\n\nWe recommend doubling the Lod Data Resolution on the OceanRenderer component to preserve detail after making this change, but note that this will "
+            + "consume 4x more video memory until we are able to optimise data usage further, so apply this change with caution.")]
+        [SerializeField, Range(1f, 4f)]
+        float _waveResolutionMultiplier = 1f;
+        public float WaveResolutionMultiplier => _waveResolutionMultiplier;
```
Comment (daleeidd) [inline crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs:31]: Should the divide-by-zero guard be applied here instead? There are a couple of lines missed below.

File: crest/Assets/Crest/Crest/Scripts/Shapes/FFT/ShapeFFT.cs [hunk 3]
```diff
@@ -161,7 +162,7 @@ public void Draw(LodDataMgr lodData, CommandBuffer buf, float weight, int isTran
                     buf.SetGlobalInt(LodDataMgr.sp_LD_SliceIndex, lodIdx);
                     buf.SetGlobalFloat(RegisterLodDataInputBase.sp_Weight, finalWeight);
                     buf.SetGlobalInt(sp_WaveBufferSliceIndex, _waveBufferSliceIndex);
-                    buf.SetGlobalFloat(sp_AverageWavelength, Wavelength * 1.5f);
+                    buf.SetGlobalFloat(sp_AverageWavelength, Wavelength * 1.5f / OceanRenderer.Instance._lodDataAnimWaves.Settings.WaveResolutionMultiplier);
```
Comment (daleeidd) [inline crest/Assets/Crest/Crest/Scripts/Shapes/FFT/ShapeFFT.cs:165]: Divide-by-zero guard.

File: crest/Assets/Crest/Crest/Scripts/Shapes/ShapeGerstner.cs [hunk 4]
```diff
@@ -137,7 +137,7 @@ public void Draw(LodDataMgr lodData, CommandBuffer buf, float weight, int isTran
                     buf.SetGlobalInt(LodDataMgr.sp_LD_SliceIndex, lodIdx);
                     buf.SetGlobalFloat(RegisterLodDataInputBase.sp_Weight, finalWeight);
                     buf.SetGlobalInt(sp_WaveBufferSliceIndex, _waveBufferSliceIndex);
-                    buf.SetGlobalFloat(sp_AverageWavelength, Wavelength * 1.5f);
+                    buf.SetGlobalFloat(sp_AverageWavelength, Wavelength * 1.5f / OceanRenderer.Instance._lodDataAnimWaves.Settings.WaveResolutionMultiplier);
```
Comment (daleeidd) [inline crest/Assets/Crest/Crest/Scripts/Shapes/ShapeGerstner.cs:140]: Divide-by-zero guard.

**Review by daleeidd — CHANGES_REQUESTED** [review]
Missed a few.

File: crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs (hunk 2 above)
Comment (huwb) [reply crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs]: D'oh - i literally had the same idea but thought "hm no its only 2 places so its not worth it". Thanks did that now

File: crest/Assets/Crest/Crest/Scripts/Shapes/FFT/ShapeFFT.cs (hunk 3 above)
Comment (huwb) [reply crest/Assets/Crest/Crest/Scripts/Shapes/FFT/ShapeFFT.cs:165]: done

File: crest/Assets/Crest/Crest/Scripts/Shapes/ShapeGerstner.cs (hunk 4 above)
Comment (huwb) [reply crest/Assets/Crest/Crest/Scripts/Shapes/ShapeGerstner.cs:140]: done - moved guard into property

**Review by daleeidd — APPROVED** [review]
Looks good. Nice one.

File: crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs (hunk 2 above)
Comment (daleeidd) [reply crest/Assets/Crest/Crest/Scripts/LodData/Settings/SimSettingsAnimatedWaves.cs]: Looks good

**Pull request merged**